### PR TITLE
Decouple JavacOptions from DefaultJavaLibrary

### PR DESCRIPTION
**Pull request by grumpyjames**
Summary:
Instead, pass an instance of JavacStepFactory. This means
that DefaultJavaLibrary::getProperties is now called only from
Project. Android extensions to DefaultJavaLibrary extend their
javacOptions at step creation time by providing a custom
JavacOptionsAmender.

Test Plan: CI

### CODE REVIEW

**Comment by grumpyjames** [comment]
Is there anything we can do to help get this reviewed? We don't want to move too far from tip, at the same time though, we're really keen to make some progress. Let me know where to send the cake ;-)

File: src/com/facebook/buck/jvm/java/JavacOptionsAmender.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+
+public interface JavacOptionsAmender extends RuleKeyAppendable {
+  JavacOptions amend(JavacOptions original, BuildContext context);
+
+  String RULE_KEY_NAME = "javacOptionsAmender";
+
+  JavacOptionsAmender IDENTITY = new JavacOptionsAmender() {
+    @Override
+    public RuleKeyBuilder appendToRuleKey(RuleKeyBuilder builder) {
+      return builder.setReflectively(RULE_KEY_NAME, "id");
```
Comment (Coneko) [inline src/com/facebook/buck/jvm/java/JavacOptionsAmender.java]: I think this shouldn't append anything to the rule key.

File: src/com/facebook/buck/android/BootClasspathAppender.java [hunk 2]
```diff
@@ -0,0 +1,37 @@
...
+import com.facebook.buck.jvm.java.JavacOptions;
+import com.facebook.buck.jvm.java.JavacOptionsAmender;
+import com.facebook.buck.rules.BuildContext;
+import com.facebook.buck.rules.RuleKeyBuilder;
+
+
+class BootClasspathAppender implements JavacOptionsAmender {
+  @Override
+  public RuleKeyBuilder appendToRuleKey(RuleKeyBuilder builder) {
+    return builder.setReflectively(JavacOptionsAmender.RULE_KEY_NAME, "android");
```
Comment (Coneko) [inline src/com/facebook/buck/android/BootClasspathAppender.java]: This shouldn't use `JavacOptionsAmender.RULE_KEY_NAME`, it should use its own key.

**Comment by Coneko** [comment]
@facebook-github-bot import

**Comment by Coneko** [comment]
Ok, just those two notes, the rest looks good.

**Comment by Coneko** [comment]
I'll fix them up then ship this, you don't need to update this unless you have the changes ready.

**Comment by grumpyjames** [comment]
That sounds great, thanks again.

**Pull request closed without merging**